protobuf_serialization/reader.nim

```diff
@@ -0,0 +1,295 @@
+#Parses the Protobuf binary wire protocol into the specified type.
+
+import options
+
+import stew/shims/macros
+import faststreams/inputs
+import serialization
+
+import internal
+import types
+
+const
+  FIELD_NUMBER_MASK: byte = 0b1111_1000
+  WIRE_TYPE_MASK: byte = 0b0000_0111
+
+type
+  ProtobufReadError* = object of ProtobufError
+  ProtobufEOFError* = object of ProtobufReadError
+  ProtobufMessageError* = object of ProtobufReadError
+
+#We don't cast this back to a ProtobufWireType despite exclusively comparing it against ProtobufWireTypes.
+#This is so an invalid wire type doesn't trigger boundChecks.
+template wireType(key: byte): byte =
+  key and WIRE_TYPE_MASK
+
+template fieldNumber(key: byte): byte =
+  (key and FIELD_NUMBER_MASK) shr 3
+
+proc handleReadException*(
+  reader: ProtobufReader,
+  Record: type,
+  fieldName: string,
+  field: auto,
+  err: ref CatchableError
+) =
+  raise err
+
+proc eofSafeRead(stream: InputStream): byte =
+  if not stream.readable():
+    raise newException(ProtobufEOFError, "Couldn't read the next byte from this stream despite expecting one.")
+  result = stream.read()
+
+#Ideally, these would be in a table.
+#That said, due to the context specific return type, which goes beyond the wire type, you need generics.
+#As Generic types aren't concrete, they can't be used in a table.
+proc readVarInt[B; E: VarIntWrapped](
+  stream: InputStream,
+  fieldVar: var B,
+  encoding: E,
+  key: byte
+) {.raises: [Defect, IOError, ProtobufEOFError, ProtobufMessageError].} =
+  type T = flatType(B)
+  when E is not VarIntWrapped:
+    {.fatal: "Tried to read a VarInt without a specified encoding. This should never happen.".}
+  when sizeof(T) == 8:
+    type
+      S = int64
+      U = uint64
+  else:
+    type
+      S = int32
+      U = uint32
+
+  if key.wireType != byte(VarInt):
+    raise newException(ProtobufMessageError, "Invalid wire type for a VarInt.")
+
+  var
+    value = U(0)
+    offset = 0'i8
+    next = VAR_INT_CONTINUATION_MASK
+    preResult: T
+  while (next and VAR_INT_CONTINUATION_MASK) != 0:
+    next = stream.eofSafeRead()
+    value += (next and U(VAR_INT_VALUE_MASK)) shl offset
+    offset += 7
+
+  #Unsigned, requiring no further work.
+  when E is UIntWrapped:
+    preResult = T(value)
+  #Zig-zagged.
+  elif E is SIntWrapped:
+    preResult = T(S(value shr 1) xor -S(value and U(0b0000_0001)))
+  else:
+    #Not zig-zagged, yet negative.
+    if offset == 70:
+      #This should handle the lowest possible negative value.
+      #The cast to a signed value causes it to error/wrap to the lowest value.
+      #Said lowest value will be negative, multiplied by -1, and wrap again.
+      #This behavior requires boundChecks to be turned off in order to not raise though.
+      {.push boundChecks: off.}
+      preResult = T(-S(value + 1))
+      {.pop.}
+    #Not zig-zagged, yet positive.
+    else:
+      preResult = T(value)
+
+  #Box the result back up.
+  box(fieldVar, preResult)
```

Comment:
If you use a reader table, like in the Json library, the copy that `box` introduces here can be avoided.

The logic will become the following:

1) Read a field ID from the stream
2) Check if we have a reader associated with the ID (the ID may be bogus)
3) If we have a reader, call it by passing the stream

Now, the reader knows that type it wants to construct and it already knows that the field is present, so it can unconditionally construct the "box" in which the final value will be placed (either a ref allocation or a `Option[T]` value). Once the box is constructed, it just calls the normal `readValue` function for the type to read the actual value from the stream.